designs/inheritance-design.md

```diff
@@ -0,0 +1,269 @@
+Expected output -> TYPE ERROR: Assigning super class to subclass
+
+# No.10 Chained calls
+```
+class CL(object):
+    def sum(self : CL) -> int:
+        return 1//0
+    def count(self : CL) -> int:
+        return 1//0
+
+class Empty(CL):
+    def sum(self : Empty) -> int:
+        return 0
+    def count(self: Empty) -> int:
+        return 0
+
+class Single(CL):
+    val : int = 0
+    def new(self : Single, val : int) -> Single:
+        self.val = val
+        return self
+    def count(self: Single) -> int:
+        return 1
+    def sum(self : Single) -> int:
+        return self.val
+    def getval(self: Single) -> int:
+        return self.val
+
+class Concat(CL):
+    left : CL = None
+    right : CL = None
+    def sum(self : Concat) -> int:
+        return self.left.sum() + self.right.sum()
+    def count(self : Concat) -> int:
+        return self.left.count() + self.right.count()
+    def new(self : Concat, l : CL, r : CL) -> Concat:
+        self.left = l
+        self.right = r
+        return self
+
+l : CL = None
+l = Concat().new(Single().new(5), Empty())
+print(l.count())
+print(Single().getval())
+```
+Expected output ->
+```
+1
+0
+```
+
+######
+
+## Design changes
+
+## AST.ts
+```
+Added supers in class like below to store the super classes for each class. It's an array because we will support
+multiple inheritance in future
+
+export type Class<A> = { a?: A, name: string, supers: Array<string>, fields: Array<VarInit<A>>, methods: Array<FunDef<A>>}
+
+```
+
+## IR.ts
+```
+Added supers in class like below to store the super classes for each class. It's an array because we will support
+multiple inheritance in future
+
+export type Class<A> = { a?: A, name: string, supers: Array<string>, fields: Array<VarInit<A>>, methods: Array<FunDef<A>>}
+
+```
+
+## type-check.ts
+```
+1) Currently we have a map from className -> [{FieldNames,Type},{FuncName,{[ParameterTypes],RetType}}]
+
+    We need to add super class info in the GlobalEnv, as below.
+
+    className -> [{FieldNames,Type},{FuncName,{[ParameterTypes],RetType}},[SuperClassType]]
+
+    classes: Map<string, [Map<string, Type>, Map<string, [Array<Type>, Type]>,Array<Type>]>
+
+2) New function isSubClass(t1:Type,t2:Type)
+    This will check if t1 is a subclass of t2
+
+   Modify isSubtype to below:
+    export function isSubtype(env: GlobalTypeEnv, t1: Type, t2: Type): boolean {
+        return equalType(t1, t2) || isSubClass(t1,t2) ||t1.tag === "none" && t2.tag === "class" 
+    }
+```
+
+## compiler.ts
+For Week 7, we do not plan to support multiple inheritance yet. Thus, our approach for implementing inherited method calls follows from Lectures 9 and 10.
+We will add a vtable and use ```call_indirect``` to reference the correct method calls.
+The compiler will need to build a ```funcref``` table and populate it with all the class-prefixed methods.
+Furthermore, each object struct will need to be augmented with a field that stores its class's offset into the table.
```

Comment:
Why do we need a field to store the offset for each "object struct"? knowing the class type should provide that information, right? can we determine this in type check?

Reply:
Yeah I guess having the offsets recorded globally based on class could be sufficient. I wrote this because I thought that was how we did it in class.
It probably makes more sense for the compiler/code generator to determine the offsets rather than the typechecker. I don't want to change the typechecker interface at the top level; the compiler builds up a global environment already for vardefs, so it makes sense to set up the order of the classes in the funcref table there too.